x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -1094,7 +1110,49 @@ private static Expression processIn(In in) {
             return childrenChanged ? in.replaceChildren(newChildren) : in;
         }
 
-        private static boolean canCastMixedNumericTypes(EsqlScalarFunction f) {
+        private static boolean processDateTimeArithmeticOperation(
+            Expression left,
+            Expression right,
+            List<Expression> newChildren,
+            String errorMessage
+        ) {
+            boolean childrenChanged = false;
+            Expression newLeft = left;
+            Expression newRight = right;
+            if (isStringLiteral(left)) {
+                newLeft = castToTemporalOrDateTime(left, errorMessage);
+            }
+            if (isStringLiteral(right)) {
+                newRight = castToTemporalOrDateTime(right, errorMessage);
+            }
+            if (newLeft != left || newRight != right) {
+                newChildren.add(newLeft);
+                newChildren.add(newRight);
+                childrenChanged = true;
+            }
+            return childrenChanged;
+        }
+
+        private static boolean isStringLiteral(Expression e) {
+            return e instanceof Literal l && l.dataType() == KEYWORD;
+        }
+
+        private static Expression castToTemporalOrDateTime(Expression e, String errorMessage) {
+            Expression result = castStringLiteralToTemporalAmount(e); // try casting it to temporal first
+            if (result instanceof Literal nr && isTemporalAmount(nr.dataType())) {
+                return result;
+            } else {
+                result = castStringLiteral(e, DATETIME); // rhs is not a temporal, try casting it to datetime
+                if (result instanceof Literal nr && nr.dataType() == DATETIME) {
+                    return result;
+                } else if (result instanceof UnresolvedAttribute ua) {
+                    return ua.withUnresolvedMessage(format(errorMessage, e.fold()));
```

Comment:
The `ua` can have a message that's relevant to the user: if i try a `datetime + "1 dayy"`, the fact that `dayy` is wrong is lost and the message we return is `Cannot convert string [1 dayy] to any of [DATETIME, DATE_PERIOD, TIME_DURATION]`.

Reply:
The code to do implicit casting for `+` and `-` will be removed.